frostsnap_coordinator/src/bitcoin/address_metadata.rs

```diff
@@ -0,0 +1,152 @@
+use crate::persist::{Persist, ToStringWrapper};
+use bdk_chain::bitcoin::bip32::DerivationPath;
+use frostsnap_core::{tweak::BitcoinBip32Path, Gist, KeyId};
+use rusqlite::params;
+use std::collections::BTreeMap;
+use tracing::{event, Level};
+
+// Presence of an address in this struct is used to identify whether
+// an address has been "shared" i.e. reserved. So that when users next
+// press receive it will show them a different address.
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+pub struct AddressMetadata {
+    // key_id, derivation, label (not yet used)
+    addresses: BTreeMap<(KeyId, DerivationPath), Option<String>>,
+}
+
+#[derive(Debug, Clone)]
+pub enum Mutation {
+    Reveal {
+        key_and_path: (KeyId, DerivationPath),
+        label: Option<String>,
+    },
+}
+
+impl Gist for Mutation {
+    fn gist(&self) -> String {
+        match self {
+            Mutation::Reveal { .. } => "Reveal",
+        }
+        .to_string()
+    }
+}
+
+impl AddressMetadata {
+    fn apply_mutation(&mut self, mutation: &Mutation) -> bool {
+        match mutation {
+            Mutation::Reveal {
+                key_and_path,
+                label,
```

Comment:
I don't think that label should be part of the `Reveal` mutation. Assigning a label is a different mutation since you can do it after an address has been revealed. I will push a commit to fix this.